awx/main/dispatch/worker/base.py

```diff
@@ -0,0 +1,163 @@
+        else:
+            queue = consumer.total_messages % len(consumer.pool)
+        return queue
+
+
+class LeastBusyStrategy(object):
+    '''
+    Attempt to deliver messages to the *least busy* worker (the worker with the
+    shortest work queue).
+    This is used for task dispatch, where we want to schedule *evenly* to run
+    as many simultaneous tasks as possible (ideally, each worker _never_ has
+    a message backlog).
+    '''
+
+    def schedule(self, consumer, body, message):
+        queue = 0
+        size = sys.maxint
+        for i, worker in enumerate(consumer.pool.workers):
+            if len(worker.managed_tasks) < size:
+                queue = i
+                size = len(worker.managed_tasks)
+        return queue
+
+
+class AWXConsumer(ConsumerMixin):
+
+    def __init__(self, name, connection, worker, queues=[], pool=None, strategy=RoundRobinStrategy):
+        self.connection = connection
+        self.total_messages = 0
+        self.queues = queues
+        self.worker = worker
+        self.pool = pool
+        self.strategy = strategy()
+        if pool is None:
+            self.pool = WorkerPool()
+        self.pool.init_workers(self.worker.work_loop)
+
+    def get_consumers(self, Consumer, channel):
+        logger.debug("Listening on {}".format(self.queues))
+        return [Consumer(queues=self.queues, accept=['json'],
+                         callbacks=[self.process_task])]
+
+    def control(self, body, message):
+        logger.warn(body)
+        if body['control'] == 'status':
+            producer = Producer(
+                channel=self.connection,
+                routing_key=message.properties['reply_to']
+            )
+            producer.publish(self.pool.debug())
+        elif body['control'] == 'reload':
+            for worker in self.pool.workers:
+                worker.quit()
+        else:
+            logger.error('unrecognized control message: {}'.format(body['control']))
+        message.ack()
+
+    def process_task(self, body, message):
+        if 'control' in body:
+            return self.control(body, message)
+        queue = self.strategy.schedule(self, body, message)
+        self.pool.write(queue, body)
+        self.total_messages += 1
+        message.ack()
+
+    def run(self, *args, **kwargs):
+        signal.signal(signal.SIGINT, self.stop)
+        signal.signal(signal.SIGTERM, self.stop)
+        self.worker.on_start()
+        super(AWXConsumer, self).run(*args, **kwargs)
+
+    def stop(self, signum, frame):
+        self.should_stop = True  # this makes the kombu mixin stop consuming
+        logger.debug('received {}, stopping'.format(signame(signum)))
+        self.worker.on_stop()
+        raise SystemExit()
+
+
+class BaseWorker(object):
+
+    def work_loop(self, queue, finished, idx, *args):
+        ppid = os.getppid()
+        signal_handler = WorkerSignalHandler()
+        while not signal_handler.kill_now:
```

Comment:
This is the main worker loop.  Every process that we spawn into a worker pool follows this logic:

1.  While `not kill_now`:
2.  Try to grab a message from my queue and call `perform_work` on that message.
3.  `perform_work` is defined in subclasses in the two worker implementations.  One of those implementations handles events (it's almost entirely the code from the 3.3.0 callback receiver).  The other implementation is the one that does what celery did in 3.3.0 - takes tasks, deserializes them, and runs Python code.